hazelcast/src/main/java/com/hazelcast/instance/DefaultNodeExtension.java

```diff
@@ -230,7 +244,39 @@ public void onClusterStateChange(ClusterState newState, boolean isTransient) {
     }
 
     @Override
+    public void onClusterVersionChange(Version newVersion) {
+        NodeEngineImpl nodeEngine = node.getNodeEngine();
+        ServiceManager serviceManager = nodeEngine.getServiceManager();
+        List<ClusterVersionListener> listeners = serviceManager.getServices(ClusterVersionListener.class);
+        for (ClusterVersionListener listener : listeners) {
+            listener.onClusterVersionChange(newVersion);
+        }
+        // also trigger cluster version change on explicitly registered listeners
+        if (!clusterVersionListeners.isEmpty()) {
+            for (ClusterVersionListener listener : clusterVersionListeners) {
+                listener.onClusterVersionChange(newVersion);
+            }
+        }
+    }
+
+    @Override
+    public boolean isNodeVersionCompatibleWith(Version clusterVersion) {
```

Comment:
Should this not be a function implemented on the version itself? 

E.g. if(node.getVersion().isCompatible(clusterVersion))...

This way you keep the NodeExtension a bit cleaner. The code becomes a bit more cohesive since the version related functions are bundled in the version.